go/libkb/secret_store_secretservice.go

```diff
@@ -58,8 +63,58 @@ func (s *SecretStoreSecretService) maybeRetrieveSingleItem(mctx MetaContext, srv
 	return &item, nil
 }
 
-func (s *SecretStoreSecretService) RetrieveSecret(mctx MetaContext, username NormalizedUsername) (secret LKSecFullSecret, err error) {
-	defer mctx.TraceTimed("SecretStoreSecretService.RetrieveSecret", func() error { return err })()
+func (s *SecretStoreRevokableSecretService) keystoreDir(mctx MetaContext, username string) string {
+	return fmt.Sprintf("ring/%s", username)
+}
+
+func (s *SecretStoreRevokableSecretService) secretlessKeystore(mctx MetaContext, username string) SecretlessErasableKVStore {
+	return NewSecretlessFileErasableKVStore(mctx, s.keystoreDir(mctx, username))
+}
+
+func (s *SecretStoreRevokableSecretService) keystoreKey() string {
+	return "key"
+}
+
+func (s *SecretStoreRevokableSecretService) keystore(mctx MetaContext, username string, keyringSecret []byte) ErasableKVStore {
+	keygen := func(mctx MetaContext, noise NoiseBytes) (xs [32]byte, err error) {
+		// hkdf with salt=nil, info=context string, and using entropy from both
+		// the noise in the file and the secret in the keyring. Thus, when we
+		// try to erase this secret, as long as we are able to delete it from
+		// either the noise file or the keyring, we'll have succeeded in making
+		// the secret impossible to retrieve.
+		// See additional docs at https://keybase.io/docs/crypto/local-key-security.
+		h := hkdf.New(sha256.New, append(noise[:], keyringSecret...), nil, []byte(DeriveReasonLinuxRevokableKeyring))
+		_, err = io.ReadFull(h, xs[:])
+		if err != nil {
+			return [32]byte{}, err
+		}
+		return xs, nil
+	}
+	return NewFileErasableKVStore(mctx, s.keystoreDir(mctx, username), keygen)
+}
+
+var identifierKeystoreSuffix = ".user"
+
+func (s *SecretStoreRevokableSecretService) identifierKeystoreKey(username NormalizedUsername) string {
+	return string(username) + identifierKeystoreSuffix
+}
+
+func (s *SecretStoreRevokableSecretService) identifierKeystore(mctx MetaContext) ErasableKVStore {
+	plaintextKeygen := func(mctx MetaContext, noise NoiseBytes) (xs [32]byte, err error) {
+		return sha256.Sum256(noise[:]), nil
+	}
+	return NewFileErasableKVStore(mctx, "ring-identifiers", plaintextKeygen)
+}
+
+func (s *SecretStoreRevokableSecretService) RetrieveSecret(mctx MetaContext, username NormalizedUsername) (secret LKSecFullSecret, err error) {
+	defer mctx.TraceTimed("SecretStoreRevokableSecretService.RetrieveSecret", func() error { return err })()
+
+	identifierKeystore := s.identifierKeystore(mctx)
```

Comment:
Ok, so let me get this straight.  There's a 1:1 mapping between EraseableKVKeystores and home directories. So we can lookup up a (homedir,username) -> (identifier) mapping without ambiguity by using this second KV store. Can we have some nice commentary about this?

BTW, remind me, how does the eraseableKVStore get erased? Is it different rules for this stuff then it is for ephemeral keys?

Reply:
Yes. Will add comments. It doesn't get erased automatically by anything, only when we `ClearSecret` which happens on logout and unclicking "remember my passphrase".